salt/_states/metalk8s_kubernetes.py

```diff
@@ -0,0 +1,170 @@
+    else:
+        # This happens if the DELETE call fails with a 404 status
+        ret['comment'] = 'The object does not exist'
+
+    return ret
+
+
+def object_present(name, manifest=None, **kwargs):
+    """Ensure that the object is present.
+
+    Arguments:
+        name (str): Path to a manifest yaml file
+                    or just a name if manifest provided
+        manifest (dict): Manifest content
+    """
+    ret = {'name': name, 'changes': {}, 'result': True, 'comment': ''}
+
+    # Only pass `name` if we have no manifest
+    name_arg = None if manifest else name
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name=name_arg, manifest=manifest, saltenv=__env__,
+        **kwargs
+    )
+    if __opts__['test']:
+        ret['result'] = None
+        ret['comment'] = 'The object is going to be {}'.format(
+            'created' if obj is None else 'replaced'
+        )
+        return ret
+
+    if obj is None:
+        __salt__['metalk8s_kubernetes.create_object'](
+            name=name_arg, manifest=manifest, saltenv=__env__,
+            **kwargs
+        )
+        ret['changes'] = {'old': 'absent', 'new': 'present'}
+        ret['comment'] = 'The object was created'
+
+        return ret
+
+    # TODO: Attempt to handle idempotency as much as possible here, we don't
+    #       want to always replace if nothing changed. Currently though, we
+    #       don't know how to achieve this, since some fields may be set by
+    #       api-server or by the user without us being able to distinguish
+    #       them.
+    new = __salt__['metalk8s_kubernetes.replace_object'](
+        name=name_arg, manifest=manifest, old_object=obj,
+        saltenv=__env__,
+        **kwargs
+    )
+    diff = __utils__['dictdiffer.recursive_diff'](obj, new)
+    ret['changes'] = diff.diffs
+    ret['comment'] = 'The object was replaced'
+
+    return ret
+
+
+def object_updated(name, manifest=None, **kwargs):
+    """Update an existing object.
+
+    Arguments:
+        name (str): Path to a manifest yaml file
+                    or just a name if manifest provided
+        manifest (dict): Manifest content
+    """
+    ret = {'name': name, 'changes': {}, 'result': True, 'comment': ''}
+
+    # Only pass `name` if we have no manifest
+    name_arg = None if manifest else name
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name=name_arg, manifest=manifest, saltenv=__env__,
+        **kwargs
+    )
+
+    cmp_manifest = manifest
+    if not cmp_manifest:
+        try:
+            cmp_manifest = __salt__[
+                'metalk8s_kubernetes.read_and_render_yaml_file'
+            ](
+                source=name,
+                template=kwargs.get('template', 'jinja'),
+                context=kwargs.get('defaults'),
+                saltenv=__env__
+            )
+        except Exception:  # pylint: disable=broad-except
+            # Do not fail if we are not able to load the YAML,
```

Comment:
Is that a wise thing to do? If the source can't be rendered, there's a programmer error, no?

Reply:
Yes but just don't want to raise here as the logic is more in the module so if the manifest really can't be loaded by the module then the module will raise and the state will fail with the right error.